File: backend/src/api/v1/graph.py
from fastapi import APIRouter
from logging import info
from src.stores.factory import get_store

router = APIRouter()


@router.get("/data")
async def get_graph_data():
    store = await get_store()
    courses = await store.get_all_courses()
    deps = await store.get_course_dependencies()

    nodes = [
        {
            "id": str(c.id),
            "label": c.title,
            "group": c.category.value,
            "title": c.description,
            "recommended_semester": c.recommended_semester,
        }
        for c in courses.values()
    ]
    info(f"Fetched {len(nodes)} courses for graph nodes: {nodes}")
    edges = [
        {
            "from": str(d.required_course_id),
            "to": str(d.course_id),
            "label": d.dependency_type.value,
        }
        for d in deps
    ]

    return {"nodes": nodes, "edges": edges}

File: backend/api/graph.go
package api

import (
	"log"
	"net/http"

	"github.com/cu-3rd-party/cu-roadmap/backend/store"
	"github.com/gin-gonic/gin"
)

func RegisterGraphRoutes(rg *gin.RouterGroup) {
	rg.GET("/data", getGraphData)
}

func getGraphData(c *gin.Context) {
	s := store.GetStore()
	if s == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": "store not initialized"})
		return
	}

	courses, err := s.GetAllCourses()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	deps, err := s.GetCourseDependencies()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	var nodes []gin.H
	for _, c := range courses {
		var recSem interface{}
		if c.RecommendedSemester != nil {
			recSem = *c.RecommendedSemester
		}
		var desc string
		if c.Description != nil {
			desc = *c.Description
		}
		nodes = append(nodes, gin.H{
			"id":                   c.ID.String(),
			"label":                c.Title,
			"group":                string(c.Category),
			"title":                desc,
			"recommended_semester": recSem,
		})
	}
	log.Printf("Fetched %d courses for graph nodes", len(nodes))

	var edges []gin.H
	for _, d := range deps {
		edges = append(edges, gin.H{
			"from":  d.RequiredCourseID.String(),
			"to":    d.CourseID.String(),
			"label": string(d.DependencyType),
		})
	}

	c.JSON(http.StatusOK, gin.H{"nodes": nodes, "edges": edges})
}
